### PR TITLE
Return transport Responses on dispatcher.Send

### PR DESCRIPTION
**Pull request by eloycoto**
This PR adds a way to get transport.Responses on the dispatcher.Send, so workers are allowed to know what happens with their Send Request.

### CODE REVIEW

**Review by strider — APPROVED** [review]
LGTM.

File: messages.go [hunk 1]
```diff
@@ -123,3 +123,25 @@ type Data struct {
...
+}
+
+type TransportResponse struct {
+	// StatusCode response
+	StatusCode int
+	// Response Body
+	Body []byte
+	// IsAsync defines if the transport is  async or sync
+	IsAsync bool
+	// Mestadata added by the transport, in case of http are the headers
```
Comment (jakub-dzon) [inline messages.go:145]: `Mestadata` -> `Metadata`

File: cmd/yggd/grpc.go [hunk 2]
```diff
@@ -16,6 +17,10 @@ import (
 	"google.golang.org/grpc"
 )
 
+var (
+	sendResponseTimeout = 20 * time.Second
```
Comment (subpop) [inline cmd/yggd/grpc.go:21]: Is this longer than the gRPC connection timeout? Or does that default to 30 seconds.

**Review by subpop — CHANGES_REQUESTED** [review]
I think you hit on a neat idea of including a channel over which to receive a response, but we need to rework the implementation some more. I'm not convinced that leaking the TransportReponse type into various packages within the `internal` package is the right solution. I'm not exactly sure at which layer this `Request`/`Response` pattern needs to end up yet. We've put a lot of responsibility into `yggd.Client`; perhaps that type needs to get cleaned up more and moved into `internal` providing more/most of the yggd functionality as an API?

File: internal/transport/transporter.go [hunk 3]
```diff
@@ -10,7 +14,7 @@ type DataReceiveHandlerFunc func([]byte, string)
 type Transporter interface {
 	Connect() error
 	Disconnect(quiesce uint)
-	SendData(data []byte, dest string) error
+	SendData(data []byte, dest string) (*yggdrasil.TransportResponse, error)
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply internal/transport/transporter.go:17]: The thing is that if we do in this way, we lost the `status_code` (And we need it), I'm ok to have another struct here similar to this. I followed the same as yggdrasil.APIResponseError in this case. 

Are you ok, if I set transport.APIResponse similar to yggdrasil one?

[the remark this answers is not in the document]
Comment (subpop) [reply internal/transport/transporter.go:17]: What about: `SendData(data []byte, dest string) (code int, resp []byte, err error)`?

Or maybe even define `code` as an enum?

```
type TransporterResponseCode int

const (
    TransporterResponseCodeSuccess TransporterResponseCode = 0
    TransporterResponseCodeFail    TransporterResponseCode = 1
)

/* ... */

SendData(data []byte, dest string) (code TransporterResponseCode, resp []byte, err error)
```

[the remark this answers is not in the document]
Comment (eloycoto) [reply internal/transport/transporter.go:17]: sorry! also the header can be needed, just because of any Aut{n,z} token inside the headers :-( Thinking on openid-connect.

[the remark this answers is not in the document]
Comment (eloycoto) [reply internal/transport/transporter.go:17]: so, SendData is a json.Marshal(*yggdrasil.TransportResponse) and return bytes, is that what you suggested? I'm happy to do that, but I do not think that makes sense, workers need to know what kind of transport the endpoint is using if wants to get the SendData response. 

But solution works for us.

[the remark this answers is not in the document]
Comment (subpop) [reply internal/transport/transporter.go:17]: Well, we've created the `transport.Transporter` interface to abstract away the details of the actual network protocol being used to transport the data. The current interface has two reflective methods in its definition: `SendData` and `ReceiveData`. Because the nature of the underlying network protocol can vary, we kept this interface very simple: "send some bytes" and "get some bytes". As you know, HTTP is a synchronous request/response protocol, while MQTT is an asynchronous publish/subscribe protocol. What we need to find is some sort of common ground here that works elegantly as an extension of the `Transporter` interface that both MQTT and HTTP as implementations of this interface can easily support.

The objective here is to extend the tx/rx nature of the `Transporter` interface to include the concept of a tx-response (and theoretically an rx-response). These responses should be common between all implementations (currently MQTT and HTTP). Does it matter to the `Transporter` interface what these responses look like? Does it matter to the caller of the interface methods what these responses look like? If the "response" value is not opaque, what content is meaningful to a caller of the `Transporter` interface methods? What sort of data would one expect back from a `SendData` or a `ReceiveData` call?

This is only half the objective. The other half is expanding the RPC protocol to relay these transport responses, but that should be considered in its own right. I would consider similar questions: what does a response to a "Send" RPC method look like? What sort of data is meaningful to the caller of the "Send" RPC method?

[the remark this answers is not in the document]
Comment (eloycoto) [reply internal/transport/transporter.go:17]: Ok, got it. I'll submit a few changes early next week.

File: cmd/yggd/grpc.go (hunk 2 above)
Comment (eloycoto) [reply cmd/yggd/grpc.go:20]: Ok, I try to get to what server is defined, by default is 120 seconds a bit excessive IMHO. 

The setting is  from here: 
https://pkg.go.dev/google.golang.org/grpc#NewServer

And the value is inherit from here: 
https://github.com/grpc/grpc-go/blob/v1.44.0/server.go#L173

So, I'm happy to do some kind of limit to 20-30  on the server side or if you want move this to 120seconds. 

Open here.

File: messages.go [hunk 4]
```diff
@@ -123,3 +123,14 @@ type Data struct {
 	Metadata   map[string]string `json:"metadata"`
 	Content    json.RawMessage   `json:"content"`
 }
+
+// DataMessage is similar to Data, but adds a channel where the response will
+// be send.
+type DataMessage struct {
```
Comment (subpop) [inline messages.go:129]: This envelope is a side effect of the dispatcher type implementation. I don't think it needs to exist here with the messages that get serialized and transmitted across the network transport. This definition could probably move to grpc.go rather than part of the public yggdrasil package API.

File: internal/transport/transporter.go [hunk 5]
```diff
@@ -1,6 +1,8 @@
 package transport
 
-import "crypto/tls"
+import (
```
Comment (subpop) [inline internal/transport/transporter.go:3]: This is an unnecessary formatting change. `go fmt` probably doesn't catch this because it's an acceptable format, but it's unncessary change considering we aren't adding or removing a package import here.

**Review by subpop — CHANGES_REQUESTED** [review]
Looks pretty good. I noticed a few small things to change, but otherwise I think we can get this merged. This has been very helpful for me while thinking through the further changes I want to make to the RPC interface, so I can't promise this code will stay in this form forever, it's been a valuable source of ideas for updating the protocol. Thank you!

File: internal/http/client.go [hunk 6]
```diff
@@ -3,15 +3,24 @@ package http
...
 
 	"git.sr.ht/~spc/go-log"
-	"github.com/redhatinsights/yggdrasil"
 )
 
+type Response struct {
+	// StatusCode response
+	StatusCode int
+	// Response Body
+	Body json.RawMessage
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply internal/http/client.go:19]: []byte when do a json.Marshal is encoded using byte64. This way is the way to encode it correctly as string.

[the remark this answers is not in the document]
Comment (eloycoto) [reply internal/http/client.go:19]: For reference: 
https://cs.opensource.google/go/go/+/refs/tags/go1.17.8:src/encoding/json/stream.go;l=260-277

[the remark this answers is not in the document]
Comment (subpop) [reply internal/http/client.go:19]: Oh neat, TIL!

**Comment by pkliczewski** [comment]
@subpop Thank you! :tada:

**Comment by eloycoto** [comment]
Thank you!!!!
